examples/vision/supervised-contrastive-learning.py

```diff
@@ -0,0 +1,234 @@
+
+[Supervised Contrastive Learning](https://arxiv.org/abs/2004.11362)
+(Prannay Khosla et al.) is a training methodology that outperforms
+plain crossentropy-supervised training on classification tasks.
+
+Essentially, training an image classification model with Supervised Contrastive
+Learning is performed in two phases:
+
+1. Training an encoder to learn to produce vector representations of input images such
+that representations of images in the same class will be more similar compared to
+representations of images in different classes.
+2. Training a classifier on top of the frozen encoder.
+
+Note that this example requires TensorFlow Addons, which you can install via:
+
+```python
+pip install tensorflow-addons
+```
+
+## Setup
+"""
+
+import tensorflow as tf
+import tensorflow_addons as tfa
+import numpy as np
+import keras 
+from keras import layers
+from keras.preprocessing.image import ImageDataGenerator
+
+"""
+## Prepare the data
+"""
+
+num_classes = 10
+input_shape = (32, 32, 3)
+
+# Load the train and test data splits
+(x_train, y_train), (x_test, y_test) = keras.datasets.cifar10.load_data()
+
+# Display shapes of train and test datasets
+print(f'x_train shape: {x_train.shape} - y_train shape: {y_train.shape}')
+print(f'x_test shape: {x_test.shape} - y_test shape: {y_test.shape}')
+
+"""## Using image data augmentation"""
+
+data_augmentation = keras.Sequential(
+  [
+    layers.experimental.preprocessing.Normalization(),
+    layers.experimental.preprocessing.RandomFlip("horizontal"),
+    layers.experimental.preprocessing.RandomRotation(0.02),
+    layers.experimental.preprocessing.RandomWidth(0.2),
+    layers.experimental.preprocessing.RandomHeight(0.2)
+  ]
+)
+
+# Setting the state of the normalization layer.
+data_augmentation.layers[0].adapt(x_train)
+
+"""
+## Build the encoder model
+
+The encoder model takes the image as input and turns it into a 2048-dimensional
+feature vector.
+"""
+
+def create_encoder():
+  densenet = keras.applications.ResNet50V2(
+    include_top=False,
+    weights=None,
+    input_shape=input_shape,
+    pooling='avg')
+
+  inputs = keras.Input(shape=input_shape)
+  augmented = data_augmentation(inputs)
+  outputs = resnet(augmented)
+  model = keras.Model(inputs=inputs, outputs=outputs, name="cifar10-encoder")
+  return model
+
+encoder = create_encoder()
+encoder.summary()
+
+learning_rate = 0.001
+batch_size = 265
+hidden_units = 512
+projection_units = 128
+num_epochs = 50
+dropout_rate = 0.5
+temperature = 0.05
+
+"""## Build the classification model
+
+The classification model adds a fully-connected layer on top of the encoder,
+plus a softmax layer with the target classes.
+"""
+
+def create_classifier(encoder, trainable=True):
+
+  for layer in encoder.layers:
+    layer.trainable = trainable
+
+  inputs = keras.Input(shape=input_shape)
+  features = encoder(inputs)
+  features = layers.Dropout(dropout_rate)(features)
+  features = layers.Dense(hidden_units, 'relu')(features)
```

Comment:
pass the activation as a keyword arg, `activation='relu'` (easier to read)

Reply:
Done